### DISCUSSION TITLE
A user-interface extension: Main-Screens

### DISCUSSION THREAD

**Discussion by lexonic** [probonopd/MiniDexed#551] [Show and tell]
Hello MiniDexed community. I really love the MiniDexed barebone project. A big thank you to all of you developers and voice designer(s), for this great project. 

Thinking on how to get easier access to the most basic functions like:

* Bank and Voice Names displayed on the same screen
* Quick selection of Banks and Voices
* Quick loading of Performances
* But not changing the existing MiniDexed menu

this is the concept I came up with: 

## UI-Main-Screen: The Concept

A new user-interface for play/perform operations. 

A long-press steps through the main screens: BANK+VOICE / PERFORMANCE / TG-Grouping&Mute

Double-clicking in any of the main screens will enter the well-known MiniDexed menu. Double-click to exit the menu.

This new UI can be disabled in `minidexed.ini`:

	UIMainEnabled=0

With this setting, MiniDexed behaves exactly as we know it: A double-click deep inside the menu will bring us back to the menu's home, and not exiting the menu.

The default screen presented after startup can be configured in `minidexed.ini` (this setting is only effective with `UIMainEnabled=1`):

	DefaultScreen=0   # -> start with classic MiniDexed menu
	DefaultScreen=1   # -> start with Bank+Voice screen
	DefaultScreen=2   # -> start with Performance screen

**The Screens:**

<pre>
                        +------------+
                        |            |      | short       | encoder     | click+turn           |
                        V            |      +-------------+-------------+----------------------+
Bank+Voice      +----------------+   |      |             |             |                      |
                |001 BankNameXY 8|   |      | SelectTG    | ChangeTG    | { MasterVolume       |
                |  1 VoiceNameX  |   |      | SelectBank  | ChangeBank  | {                    |
                +----------------+   |      | SelectVoice | ChangeVoice | {                    |
                        | long       |      |             |             |                      |
                        V            |      +-------------+-------------+----------------------+
Performance     +----------------+   |      |             |             |                      |
                |PERF 000-001 (L)|   |      | LoadPerf    | ChangePerf  | MasterVolume         |
                |PerformancName  |   |      |             |             |                      |
                +----------------+   |      |             |             |                      |
                        | long       |      |             |             |                      |
                        V            |      +-------------+-------------+----------------------+
Groups+Mute     +----------------+   |      |             |             |                      |
                |Group AABBC..C  |   |      | mute/unmute | SelectTGNum | EditGroup (none,A-D) |
                |TG    12345678  |   |      |             |             |                      |
                +----------------+   |      |             |             |                      |
                        | long       |      +-------------+-------------+----------------------+
                        |            |
                        +------------+      double-click = enter menu

MasterVolume    +----------------+      pop-up screen during manual MasterVolume change 
                |Master Volume   |
                |[███████    ]  75|
                +----------------+
</pre>

## Bank + Voice Screen

Banks and Voices can be selected for each TG.

<pre>
                           +----------------+
Bank Number + Bank Name -> |128 BankNameXY 8| <- TG-number
Voice Num. + Voice Name -> | 32 VoiceNameX ♪| <- MIDI activity indicator (not yet implemented)
                           +----------------+

                           +----------------+
                           |001[BankNameXY]8| <- [] = selected item can be changed with encoder
                           |  1 VoiceNameX ♪|
                           +----------------+
</pre>

Disadvantage of this screen layout: bank numbers are limited to 3 digits (1..999).

The selected Bank or Voice will show the name in squared brackets "[ ]". No squared brackets are visible when TG is selected. Select next item with an encoder-click. The selected item can be changed by turning the encoder.

## Performance Screen

Select and load a performance.

<pre>
                           +----------------+
                           |PERF 0001    (L)| <- (L) for loaded as in classic menu
                           |PerformanceName | <- Performance names (up to 16 characters possible)
                           +----------------+
</pre>

Select performance with encoder and load it with encoder-click.

## TG Grouping and Mute

Grouping links TG's together. Grouped TG's syncronize changes of Voices and Banks in the Bank+Voice screen. Changes in the classic menu are not syncronized yet!

Select the TG by turning the encoder. A click toggles mute/unmute. Click+turn changes the groupID.

Load/save group settings in performance files for each TG:

	TGGrouping#=0..4   # -> 0 = ungrouped; 1..4 = A..D

Muted TG's are a working tool and are not saved in the performance file. (Like disabled OP's in the DX7)

<pre>
                           +----------------+
                           |Group ..BBC..C  | <- TG's without grouping show as "."
                           |TG    1-3--678  | <- muted TGs show as "-"
                           +----------------+
</pre>

Un-grouped TG's shows as ".".

Muted TG's show as "-" instead of the TG-number.

Behaviour of grouping for TG's in the same group:

* changing a bank will be reflected on all TG's with same group ID.
* changing a voice will change to the same Voice and Bank on all TG's with same group ID.

## Master Volume Screen

MasterVolume is adjusted by encoder click+turn.

<pre>
                           +----------------+
                           |Master Volume   |
                           |[███████    ]  75| <- MasterVolume is changed in steps of 5 (from 0 to 100)
                           +----------------+
</pre>

MasterVolume-screen is not shown or updated by changing MasterVolume by MIDI.

Load global MasterVolume from `minidexed.ini` (range: 0..100):

	MasterVolume=100

## Current code for testing

The current branch of code is here: <https://github.com/lexonic/MiniDexed/tree/ui-main-screen>

The file `src/performance.ini.test` can be used for testing the grouping and muting features.

What are your thoughts about this UI extension? Is this something of interest?

REMARK: I am completely new to GitHub, and this is my first time trying to contribute to a project in GitHub. My commit history is a bit a mess and the commit messages are not that exact. I'm working to improve on this :wink:

**Comment by Banana71** [comment]
Hello Bernhard, I'm happy. I would like to try out your changes but unfortunately I can't generate the files to copy them to the SD card. Can someone please provide me with a working version or explain to me how to create the files.

**Reply by lexonic** [reply]
Hello Peter,
I tried to run the github workflow to build for all RPi versions, but yet could not find out how to get it started. I had hoped it's auto-created by doing a release, which is not :grin:

Here is a working build for RPi3:
[MiniDexed_2023-10-18.zip](https://github.com/probonopd/MiniDexed/files/13047888/MiniDexed_2023-10-18.zip)

If you need a build for another RPi, please let me know and I create it manually.

**Reply by lexonic** [reply]
A build for RPi4:
[MiniDexed_2023-10-19_RPi4.zip](https://github.com/probonopd/MiniDexed/files/13048188/MiniDexed_2023-10-19_RPi4.zip)

Serial MIDI is somehow not working with my RPi4, where the same interface works on my RPi3 A+. Couldn't find if there's an issue with this build or another problem :confused:. Nevertheless, USB MIDI is working on both.

**Reply by lexonic** [reply]
Addendum to my troubles with the RPi4 buid: To trace down the issue, I pushed in an SD loaded with the continuous build from the main branch, and serial MIDI worked. To cross-check I put back my RPi4 build. Serial MIDI worked now too! 
Maybe a broken jumper wire? But yesterday I have re-seated all connections and rebooted several times. Serial MIDI didn't work a single time. Strange things happen... :grin:

**Reply by Banana71** [reply]
> To cross-check I put back my RPi4 build. Serial MIDI worked now too!

- [x] My RPi 4 with classic MIDI input also works perfectly :star_struck:

**Reply by Banana71** [reply]
Hi Bernhard, 
Internally in the program the performance ID starts at 0 but in the display the first performance should start with 1. 
I'm not sure if this is the right place in the program, but can you add 1 to the value in the line?

uimain.cpp
526 `ValuePadded += pUIMain->PadString(to_string(nValue).c_str(), 4, '0', true);`

**Reply by lexonic** [reply]
Is there a consent on how this behavior should be? 

In previous releases, the first file in the performance directory was _000001_PerfNameWhatever.ini_. Then the _performance.ini_ from the root-directory would become "0000 Default", the rest start with number 1 matching the numbering in the performance directory. I was happily excepting 0 as primary performance and whatever is in performance directory, gets re-numbered consecutively and starting with 1.

I have realized that in your recent performance updates (#414) you are now starting with 2 (e.g. _000002_Concert D.ini_). First I thought this was random, then I have seen that in your sound-list you mention "001_Default". Is this the the new expected behavior? 

If this is common consent, I am happy to make a quick-fix to add 1 to the value.

**Reply by lexonic** [reply]
There is a branch "print-performance-number" (#531) which took the numbers from the filenames instead of renumbering consecutively, but this branch wasn't merged back into "main". What was the reason that this branch was abandoned?

**Reply by Banana71** [reply]
If in doubt, we refer to the models Yamaha DX7, DX7II, TX802,...
At that time there was no Yamaha synthesizer in which the numbering of the Voices or Performances began with 0.

> I have realized that in your recent performance updates (#414) you are now starting with 2 (e.g. _000002_Concert D.ini_). First I thought this was random, then I have seen that in your sound-list you mention "001_Default". Is this the the new expected behavior?
If this is common consent, I am happy to make a quick-fix to add 1 to the value.

This is actually just an attempt to make the sound lists match the internal numbering. However, the performance list still has gaps and it doesn't really fit together yet.
We have already had some discussions here about the numbering and in the future the numbering should also be taken over from the file name of Performance.ini.

⌛ Don't rush, it's not like the sounds won't be found again.

**Reply by Banana71** [reply]
>There is a branch "print-performance-number" (https://github.com/probonopd/MiniDexed/pull/531) which took the numbers from the filenames instead of renumbering consecutively, but this branch wasn't merged back into "main". What was the reason that this branch was abandoned?

I'm not sure anymore but I think it was the performance ID that was displayed and not the numbering of the file name.
That's why I now always create the performance list, which shows consecutive numbering and the numbering of the file name.

**Reply by lexonic** [reply]
Me personally, I also prefer having the number taken form the file-name.

BTW, I have the build of "print-performance-number" running right now with your performance set from Sept. 18. It is taking the numbers from the files. This is what i got:
```
 performance
2=Concert D
...
97=Synth Bass
100=SlaM Kit
...

```
Just working on a build script to have one SDcard for RPi2, 3, and 4. Once this works, I'll do a local build with "+1" and upload it here.

**Reply by lexonic** [reply]
Here it is: Quick-fix "Performance value +1" for RPi2 (untested), RPi3, and RPi4
[MiniDexed_2023-10-21_Perf+1_fix.zip](https://github.com/probonopd/MiniDexed/files/13061818/MiniDexed_2023-10-21_Perf%2B1_fix.zip)

BTW, I love your "Concert D", very sweet dynamics, lovely pianissimo, incredible.

**Comment by Banana71** [comment]
I tried the version for the RPi 3 with classic MIDI-Input.
My miniDexed behaves exactly as you described and I like it. :-)

My thoughts and questions:
- MasterVolume: Is the adjustable visible range (0 -100) scaled to 0 - 127 (0 - FF)?
I would save the master volume in Performance.ini.
- TG Grouping and Mute:
Unfortunately I can't see where the next change will take place. I read something about the Cousor in the code (uimain.cpp, from line 560), but no Cousor is shown on the display (SSD1306).
Grouped editing of voices would be great :-)

- The configurable DefaultScreen is a great solution.
- For everyone who thinks "Früher war alles besser"   `UIMainEnabled=0` is great. 

I will keep your version on the miniDexed
Thank You

**Reply by lexonic** [reply]
Thanks for your input.
- The number of MasterVolume shown is in percent. In MiniDexed internally it is represented as floating-point with a range of 0 to 1. Seems to be used as multiplier of the raw (unamplified) signal.   
Implementing a parameter in the global section of Performance.ini can be done, shall have priority over the MasterVolume setting from minidexed.ini.
- You are right, on LCD's there is a cursor in the "TG Grouping & Mute". I have now tested it with an SSD1306 and no cursor there.  Maybe cursors are not implemented in the circle ssd1306 library? Does it support escape-sequences? This needs some investigation. Does anyone know?

**Reply by Banana71** [reply]
>You are right, on LCD's there is a cursor in the "TG Grouping & Mute". I have now tested it with an SSD1306 and no cursor there. Maybe cursors are not implemented in the circle ssd1306 library? Does it support escape-sequences? This needs some investigation. Does anyone know?

I don't have an answer to your questions either, but perhaps it helps that there is an acceptable cusor function (works with SSD1306) in the PAN menu.

### uimenu.cpp
```
1016 string CUIMenu::ToPan (int nValue)
1017 {
1018       assert (CConfig::LCDColumns == 16);
1019       static const size_t MaxChars = CConfig::LCDColumns-3;
1020        char PanMarker[MaxChars+1] = "......:......";
1021       unsigned nIndex = nValue * MaxChars / 127;
1022        if (nIndex == MaxChars)
1023       {
1024                nIndex--;
1025        }
1026        PanMarker[nIndex] = '\xFF';                     // 0xFF is the block character
1027
1028        return PanMarker;
1029 }

```

**Reply by lexonic** [reply]
A blinking block character would be a nice solution for a cursor. Unfortunately, I could not find an easy way to make it blink (alternating with the underlying character), and I didn't want to overwrite the group or mute information on the display with a solid block.

But maybe the solution below is suitable for you?

**Comment by diyelectromusic** [comment]
Interesting idea.  I haven't looked closely at your changes, but in terms of your description, things like back/home/select are all configurable with either short/long presses or various buttons and so on - does this scheme reflect those configuration choices or does it rely on the double-click and specific long-presses to activate it?

Kevin

**Reply by lexonic** [reply]
I think I got your point. Maybe my description was not very clear. The user interface is built to work with the clickable rotary encoder (e.g. KY-040); I did not test the usability with buttons only. And no additional buttons needed. 

The functions of the various button actions (like short, long, double-click) are not configurable. Most button actions are consistent through all screens, but few change depending on the selected main-screen.

These are the actions and their functions within the main-screen:

Consistent in all main-screens:
- long press: go to the next screen. rotates through: Bank+Voice -> Performance -> TG-Group&Mute
- turning the encoder: changing the value of the selected item
- double-click: enter/exit the known MiniDexed menu. Controls inside the menu are unchanged, as we know them. Except the double-click, which used to bring us to the home position of the MiniDexed menu will now EXIT the menu and return to the main-screen.

And those vary based on the screen:
- short press: 
  - Voice+Bank: select next item
  - Performance: load selected performance
  - TG-Group&Mute: mute/unmute
- click+turn: 
  - Voice+Bank and Performance: adjust MasterVolume
  - TG-Group&Mute: select group-ID

I believe it is less complicated than it sounds, once you have your hands-on. 

Hopefully my attempt of explanation did not seed more confusion :smiley:

**Reply by diyelectromusic** [reply]
I think I understand, I guess my concern is more that it breaks the idea of configurability of the menu system controls by having fixed modes and configurations.

Whilst a really interesting approach if integrated then I'd now also be very wary about all the duplicated code in the UI.  If people want this we might need to think how to refactor things a bit better and integrate it in a more supportable way for the future.

Kevin

**Reply by diyelectromusic** [reply]
Hmm.  Looking at your code I think you are just using the standard prev/next/back/home/select events?  If so then they probably are configurable, just maybe not very accurately named for what you want to do.

We'd have to think if different buttons/events ought to be used so that the minidexed.ini entries aren't too confusing for this mode.  Again, something that needs some proper thought I guess as to how the two modes interact config wise.

Kevin

**Reply by lexonic** [reply]
Now I understand what you tried to tell me in your first post. Yes, fully agree, by using the standard BtnEvents, the names will be misleading for what I want them to do. Thanks for pointing that out.

And yes, _uimain_ contains duplicated code from _uimenu_, which definitely needs refactoring before integration. I didn't want to get too deep into modifying the existing _uimenu_ code. My initial goal was to see if the UI concept is usable and practical.

Let's see how others think of this concept and if its worth to put more thoughts into it. 

Bernhard

**Comment by plamikcho** [comment]
Tried the build for RPi3 - this is a great feature, exactly what I needed. I like the configurable default screen and also the performance number, now I can navigate faster to distant performances. Tried the TG grouping but without some kind of indication what is the current active item (a blinking dot/dash/number/letter) editing was tricky. I have SSD1306. Bank + voice is also great to be in a single screen, changing is faster.

**Comment by lexonic** [comment]
**A quick-fix for the ones who use SSD1306 displays:** 

Unfortunately, the SSD1306 device library does not support a cursor. Also a bold mode or reverse video mode is not implemented.

Maybe this is a convenient solution: in the TG-Group&Mute screen, the selected TG is now displayed on the upper-right corner.
[MiniDexed_2023-10-23.zip](https://github.com/probonopd/MiniDexed/files/13079045/MiniDexed_2023-10-23.zip)

**Reply by plamikcho** [reply]
It works and for me is perfectly fine. I am using your build since you published it. I hope it will be integrated into the main project branch, it is really great UI extension. Thank you for your work, it is appreciated.

**Reply by lexonic** [reply]
Hello Plamen, thank you for your feedback. Happy to hear that you like the UI.
I will continue to work on it to make it compliant for a possible integration into the main branch.

**Comment by probonopd** [comment]
Hi @lexonic, thanks for taking the time to improve MiniDexed. :+1: 

I think it'd be incredibly helpful ot have a video showing how to use this interface. I'd especially interested in how you use it to speed up your workflow, and how you are using the grouping feature.

To consider this for the main branch, we'd also need to overhaul the documentation in the wiki, and we need to keep in mind that newcomers (coming from e.g., Hackaday or the Floyd Steinberg videos) would find a different interface from what is described there.

**Reply by lexonic** [reply]
Hello @probonopd, thank you for your input.

I had planned the grouping feature for working with stacked voices, where the same voice is used with different parameters such as panning or detune. That's why grouping right now only affects banks and voices. But it seems that others in this community have different ideas on how to use it, so let's see what is the best way to implement it.

You are absolutely right that being consistent with the documentation and with how MiniDexed is announced is important. Right now there is a config setting in _minidexed.ini_, where users can decide if they want to work with the classic interface or the new one. We can set the classic interface as default.

Once we get close to merging the new UI into the main branch, I will work on the wiki for the new interface. But unfortunately, November will get quite busy for me, and I am afraid that I will not find much time for coding this month.

**Comment by Piattica** [comment]
About the menu extension I think it would be a very good improvement of the usability of the little fantastic MiniDexed.
IMHO the **live use of MINIDEXED** will be easier if there would be a **direct control of volume of current performance** because is always difficult to have all the performances with the same output level.

**Reply by Banana71** [reply]
>I would save the master volume in Performance.ini.

**Reply by Piattica** [reply]
my goal is to have a pararmeter that you can adjust with potentiometer while you play

**Reply by probonopd** [reply]
We are using a rotary encoder rather than a potientometer, but you can adjust all parameters with it. Having more than one rotary encoder is outside of the scope of MiniDexed, but you could make additional rotary encoders or potentiometers attached to, e.g., a Raspberry Pi Zero with some code that would send MIDI commands to MiniDexed. That way, only the imagination is the limit.

**Reply by Piattica** [reply]
sorry, my mistake; I mean rotary encoder
